src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,407 @@
+
+        /**
+         * 
+         */
+        private static final long serialVersionUID = -7259431101547222511L;
+
+        public TerminationException(String message, Throwable cause) {
+            super(message, cause);
+        }
+
+        public TerminationException(String message) {
+            super(message);
+        }
+        
+    }
+
+    private void terminateContainer(DockerCloud dc, Container container) throws TerminationException {
+        String nodeName = String.format("terminateSlave-%d", terminateNodeNameCounter++);
+        String containerId = container.getId();
+        
+        DockerClient client = dc.getDockerApi().getClient();
+        Map<String, String> containerLabels = null;
+        try {
+            containerLabels = getLabelsOfContainer(client, containerId);
+        } finally {
+            try {
+                client.close();
+            } catch (IOException e) {
+                LOGGER.info("Unable to close Docker client while trying to gracefully terminate container", e);
+            }
+        }
+        String templateName = containerLabels.get(DockerTemplate.CONTAINER_LABEL_TEMPLATE_NAME);
+        
+        // NB: dc.getTemplate(String) does not work here, as it compares images!
+        DockerTemplate template = null;
+        for (DockerTemplate dockerTemplate : dc.getTemplates()) {
+            if (dockerTemplate.getName().equals(templateName)) {
+                template = dockerTemplate;
+                break;
+            }
+        }
+        
+        if (template == null) {
+            throw new TerminationException(String.format("Template %s in DockerCloud %s does not exist", templateName, dc.toString()));
+        }
+        
+        DockerTransientNode dtn;
+        try {
+            dtn = this.createDockerTransientNode(nodeName, containerId, template.getRemoteFs());
+        } catch (FormException | IOException e) {
+            throw new TerminationException("Creating DockerTransientNode failed due to exception", e);
+        }
+        dtn.setNodeDescription(String.format("Watchdog is terminating detached Container %s", containerId));
+        dtn.setAcceptingTasks(false);
+        dtn.setDockerAPI(dc.getDockerApi());
+        dtn.setRemoveVolumes(template.isRemoveVolumes());
+        dtn.setMode(Node.Mode.EXCLUSIVE); // restrict usage as much as possible
+        dtn.setNumExecutors(0);
+        
+        dtn.terminate(LOGGER);
+        
+        LOGGER.info("Successfully terminated container {} consistently", containerId);
+    }
+
+    private void checkForSuperfluousComputer() {
+        for (Node node : nodeMap.values()) {
+            if (! (node instanceof DockerTransientNode)) {
+                // this slave does not belong to us
+                continue;
+            }
+            
+            DockerTransientNode dtn = (DockerTransientNode) node;
+            
+            Container container = this.csmMerged.getContainerById(dtn.getContainerId());
```

Comment:
Internal remark to myself: Why not using the fact that the `dtn` knows the `CloudId()` already? `CloudId` and `containerId` should then be united --- and could be concat'ed easily in the map (though still staying as two separate parameters in the prototype of the methods). We only get troubles with returning two values in Java...

Reply:
After several more rounds of thinking about it, I came to the very same conclusion as you depicted above: We don't need to know exactly which `DockerCloud` had created the `DockerTransientNode`.
That is why with 706dbe80b5b8589bfc43eabcfaccf66d071a546c I have now followed your suggestion and introduced a `ContainerNodeNameMapping.isContainerIdRegistered()`, which merely just tells us "that there is a suspect around". Thus, our coding should be very conservative and don't kill anything, which might still be required. Moreover, I have spend a longer comment at the point of usage...

What do you think?

Reply:
I've read the comment and agree that we don't need to know which cloud created the node ... but I also think we don't _need_ to know which template it was created from either.  It would be nice to know if we should also remove a container's volumes as well, but we can still do useful tidy-up work without knowing that for sure.